DM-47325: Add API for parsing butler dataset URIs (butler and ivo)

timj opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

timj committed:
Add API for parsing butler dataset URIs (butler and ivo)

timj commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -525,6 +528,83 @@ def get_known_repos(cls) -> set[str]:
+        return label, dataset_id
+
+    @classmethod
+    def get_dataset_from_uri(cls, uri: str) -> DatasetRef | None:
```
@dhirving I was thinking that this should take a `LabeledButlerFactory` as an optional parameter (so that a service like vo-cutouts can call it directly with the URI) but do you have any thoughts on how to pass in the token? In the general case the caller won't know which butler the URI is referring to. Can I assume that `**kwargs` should be forwarded as `factory.create_butler(label=label, **kwargs)` and assume the token would be shared? Or should there be a dict indexed by label that has the additional parameters that the service can set up once and pass in?

dhirving replied:
I like your idea of a config object that could be set up by the caller and passed through.

One other idea would be a dependency injection pattern.  Maybe declare a protocol like
```
class ButlerFactoryProtocol(Protocol):
    create(label: str) -> Butler

# (or just `Callable[[str], Butler]
```
which `get_dataset_from_uri` could take as a parameter.  Then `LabeledButlerFactory` could grow a method `bind(access_token: str) -> ButlerFactoryProtocol` to create a factory with the access token already bound in.  That would let callers do whatever they want to customize the instance creation.

dhirving replied:
At the moment the access token is in fact shared since it comes from Gafaelfawr and just represents the user's identity.  Though no guarantees it will stay they way if we end up needing to delegate authentication between IDF and USDF, or some other unfortunate thing.

timj replied:
Trying to think through what you said above. For:

```python
my_factory = factory.bind(access_token=token) 
ref = Butler.get_dataset_from_uri(dataset_uri, my_factory)
```
doesn't this mean that the access_token is required to be the same for every butler label?

Don't we need:

```python
my_factory = factory.bind("dp02": {"access_token": token}, "dp02-direct": {"access_token": None})
```
?

dhirving replied:
There is only one access token per HTTP request -- Gafaelfawr generates it to identify the user.  It is not associated with the repository in any way.  So at least at the moment there's only one access token.  If you pass an access token to `LabeledButlerFactory` for a DirectButler label it just ignores it.

The advantage to the dependency injection pattern is that this code wouldn't have to know anything at all about authentication or where Butlers come from -- it makes that the caller's responsibility.  This code just knows that if it has a label, the caller will work out how to get it the matching Butler.  So if things get more complicated later, this interface doesn't have to change.  (One example might be caching the Butler if someone wants to call `get_dataset_from_uri` multiple times in the same request -- though that might be better solved by wrapping this function in a separate object.)

Your config object idea would work well too, allowing us to add more options later if `LabeledButlerFactory` gets more complicated.  It's just a little less flexible and it couples this function tightly to `LabeledButlerFactory`.

timj commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -525,6 +528,83 @@ def get_known_repos(cls) -> set[str]:
+        can include ``/`` and the leading ``/`` is always stripped. If the
+        repository label starts with ``/`` then it must be doubled up. e.g.,
+
+            ivo://rubin//repo/main/82d79caa-0823-4300-9874-67b737367ee0
```
@gpdf do you have any thoughts as to what the netloc field for our IVO IDs might look like?

gpdf replied:
Following discussion with @timj today, the suggestion is that the netloc be treated leniently -- it will have a standard value (e.g., "rubin", still to be confirmed with the IVOA), and then the path will be:

butler/(butler-identity)/(uuid)

If the *butler-identity* begins with a "/", e.g., "/repo/main", then an extra "/" is required in the pathname, e.g., 

ivo://rubin/butler//repo/main/82d79caa-0823-4300-9874-67b737367ee0

To be explicit the *butler-identity* is everything between the first and last slashes in the path.

timj replied:
After discussing this with Markus it was noted that the variable part of the IVO ID has to be a query parameter. ie something like

```
ivo://lsst.org/butler?/repo/main/82d79caa-0823-4300-9874-67b737367ee0
```
or I suppose:
```
ivo://lsst.org/butler//repo/main?82d79caa-0823-4300-9874-67b737367ee0
```
Depending on whether we have a look up service per butler concept or per butler repository.
Maybe it's worth pondering what it would look like for an Object/Source ID.
```
ivo://lsst.org/object?1234567890
```
where in theory we can work out the data release number from the object ID. Or
```
ivo://lsst.org/object/dr1?1234567890
```

timj replied:
or even

```
ivo://lsst.org/dr1?1234567890
```
and
```
ivo://lsst.org/dr1?butler/82d79caa-0823-4300-9874-67b737367ee0
```

timj replied:
I am working through ideas for IVO IDs on [DMTN-302](https://dmtn-302.lsst.io).

timj replied:
For completeness we now parse `ivo://org.rubinobs/usdac/lsst-dr1?repo=dr1id=UUID`.

The parser does not attempt to look at anything other than the query component.

gpdf reviewed (changes requested):
Responded to request for feedback on ivo: URIs.

timj committed:
Add support for using a LabeledButlerFactory to retrieve the butler

timj commented on tests/test_simpleButler.py:
```diff
@@ -882,10 +883,61 @@ def makeButler(self, writeable: bool = False) -> Butler:
+            index_file.flush()
+            with mock_env({"DAF_BUTLER_REPOSITORY_INDEX": index_file.name}):
+                butler_factory = LabeledButlerFactory()
+                factory = butler_factory.bind(access_token=None)
```
@rra does this approach work for you in the cutout service?

```python
factory = butler_factory.bind(access_token=token)
...
ref = Butler.get_dataset_from_uri(dataset_uri, factory=factory)
```
?

rra replied:
The cutout service passes a Butler instance into the backend, so the pattern looks like:
```python
butler_factory = LabeledButlerFactory()

def _get_backend(label: str, token: str) -> ImageCutoutBackend:
    # Called for each cutout
    factory = butler_factory.bind(access_token=token)
    butler = factory.create_butler(label=label)
    # ...
    return ImageCutoutBackend(butler, projection_finder, output, tmpdir)
```
Is that what you had in mind? Basically move the access token parameter to `create_butler` to an intermediate step to create a factory with a bound token? If so, that would be fine here.

timj replied:
Ah. No. It's probably the wrong API for you. Somewhere you are parsing the `butler://` URI and I'm trying to provide code here that will hide the URI structure from you (so that we can also support the new `ivo://` URIs). This PR creates two APIs: one that parses the URI and returns the butler repo label and the UUID, and another API (that is the one I talk about here) that lets you retrieve the DatasetRef directly from the URI and a butler factory. Maybe `get_dataset_from_uri` should return the `Butler` instance along with the `DatasetRef`?

rra replied:
Maybe I could pass the return value of `bind` into the backend along with the URIs as-is without parsing them and then the backend can do whatever it needs to do? That would be even more convenient. In other words, have the constructor of `ImageCutoutBackend` take a Butler factor instead of a Butler instance.

timj replied:
I think passing the bound factory and URI into whatever is wanting to know the DatasetRef is what we want here.

timj committed:
Switch IVOID to using a query string for the dataset UUID

timj committed:
Change det_dataset_from_uri so it returns a Butler

Otherwise there is no easy way for the caller to get the
dataset:

butler, ref = Butler.get_dataset_from_uri(uri)
thing = butler.get(ref)

timj committed:
Change IVO parser to use from described in DMTN-302

timj committed:
Add news fragment

stvoutsin commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -526,6 +529,107 @@ def get_known_repos(cls) -> set[str]:
+        in the ``ivo`` URI query component.
+        """
+        parsed = urllib.parse.urlparse(uri)
+        if parsed.scheme == "ivo":
```
```suggestion
        scheme_lower = parsed.scheme.lower()
        if scheme_lower == "ivo":
```

stvoutsin replied:
I think for the IVOIDs the registry part (everything before ? or #) needs to be case insensitive

stvoutsin commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -526,6 +529,107 @@ def get_known_repos(cls) -> set[str]:
+                raise ValueError(f"Butler IVOID only supports a single value of repo and id, got {uri}")
+            label = qs["repo"][0]
+            id_ = qs["id"][0]
+        elif parsed.scheme == "butler":
```
```suggestion
        elif scheme_lower == "butler":
```

stvoutsin commented on python/lsst/daf/butler/_labeled_butler_factory.py:
```diff
@@ -83,6 +90,12 @@ def __init__(self, repositories: Mapping[str, str] | None = None) -> None:
         # This may be overridden by unit tests.
         self._preload_direct_butler_cache = True
 
+    def bind(self, access_token: str | None) -> LabeledButlerFactoryProtocol:
```
Perhaps a docstring here for clarity?

```suggestion
    def bind(self, access_token: str | None) -> LabeledButlerFactoryProtocol:
        """Create a callable factory function for generating Butler instances.

        Parameters
        ----------
        access_token : str | None
            An optional access token to authenticate with the Butler.

        Returns
        -------
        LabeledButlerFactoryProtocol
            A callable that takes a label as input and returns a Butler
            instance.
        """
```

timj replied:
Thanks. I'm confused as to how it is that my numpydoc linter didn't complain.

stvoutsin commented on tests/test_simpleButler.py:
```diff
@@ -882,10 +883,75 @@ def makeButler(self, writeable: bool = False) -> Butler:
+            "ivo://org.rubinobs/butler?repo=dr1&repo=dr2&id=82d79caa-0823-4300-9874-67b737367ee0",  # 2 vals.
+            "ivo://org.rubinobs/something?repo=%20&id=82d79caa-0823-4300-9874-67b737367ee0",  # no repo.
+            "https://something.edu/1234",  # Wrong scheme.
+        ):
```
Might it also be useful to test percent-encoded params?

timj replied:
The `%20` is converted to a space for the failing test here but I added a `%2F` for a `/` in the URIs that work.

stvoutsin commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -526,6 +529,107 @@ def get_known_repos(cls) -> set[str]:
+    @classmethod
+    def get_dataset_from_uri(
+        cls, uri: str, factory: LabeledButlerFactoryProtocol | None = None
+    ) -> tuple[Butler, DatasetRef | None]:
```
I think this is fine if the result tuple is unlikely to change in the future, If you think it may though, I'd also consider a dataclass/named tuple result which is a bit cleaner to change and generally more explicit.

timj replied:
I will think about it. I can't currently think of any information that could be returned given the limited content of the URI itself.

timj replied:
Switched to data class.

stvoutsin commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -526,6 +529,107 @@ def get_known_repos(cls) -> set[str]:
+        if parsed.scheme == "ivo":
+            # Do not validate the netloc or the path values.
+            qs = urllib.parse.parse_qs(parsed.query)
+            if "repo" not in qs or "id" not in qs:
```
I'm not sure what is expected of query param keys in terms of case sensitivity. But I think no harm in treating those as case-insensitive as well?

timj replied:
In theory we are generating these IVO IDs so I don't really want the added complication of converting the dict to a dict with case insensitive keys (I have a recollection of using such a special dict at some point but I'm not sure where it was).

stvoutsin reviewed (approved):
Overall looks good to me, I think the new methods make sense and are useful. Only a couple thoughts/suggestions here

timj committed:
Allow IVO:// and BUTLER:// URI schemes to be used

The URI scheme and netloc are meant to be case insensitive.
Since the butler URI scheme relies on butler labels which
are not case-insensitive, the netloc is not downcased in
that situation.

timj committed:
Ensure that URI parsing test unquotes

The %2F becomes a /

timj committed:
Add docstring for bind method

timj committed:
Return dataclasses rather than tuple

gpdf reviewed (approved):
I'm fine with this now, though we are still discussion what will go into the `authority` field in the IVOID versus what will go in the `resource key` field (hostname and path fields, colloquially speaking).

I am assured that that's not going to require additional work on the code affected by this PR.
